File: java/Assembler.java
import java.io.*;
import java.util.*;

public class Assembler {

	private static final Map<String, Integer> codes = new HashMap<>();
	private static final Map<Integer, String> codes_inv = new HashMap<>();
	private static final Map<String, Integer> lookup = new HashMap<>();
	private static final Map<String, Integer> symtbl = new TreeMap<>();

	static {
		initialize_maps();
	}

	private static void initialize_maps() {
		codes.put("and", 0x0000); codes.put("or", 0x0400); codes.put("xor", 0x0600);
		codes.put("add", 0x0800); codes.put("sub", 0x0a00); codes.put("lsl", 0x3000);
		codes.put("lsr", 0x3400); codes.put("asr", 0x3600); codes.put("ldw", 0x4000);
		codes.put("ldb", 0x4400); codes.put("lbu", 0x4600); codes.put("stw", 0x4800);
		codes.put("stb", 0x4c00); codes.put("blt", 0x8000); codes.put("bge", 0x8200);
		codes.put("bbl", 0x8800); codes.put("bae", 0x8a00); codes.put("beq", 0x9000);
		codes.put("bne", 0x9200); codes.put("tlt", 0xa000); codes.put("tge", 0xa200);
		codes.put("tbl", 0xa800); codes.put("tae", 0xaa00); codes.put("teq", 0xb000);
		codes.put("tne", 0xb200); codes.put("hlt", 0xf800);

		for (Map.Entry<String, Integer> entry : codes.entrySet()) {
			codes_inv.put(entry.getValue(), entry.getKey());
		}

		lookup.put("r0", 0); lookup.put("r1", 1); lookup.put("r2", 2); lookup.put("r3", 3);
		lookup.put("r4", 4); lookup.put("r5", 5); lookup.put("r6", 6); lookup.put("r7", 7);
		lookup.put("r8", 8); lookup.put("r9", 9); lookup.put("r10", 10); lookup.put("r11", 11);
		lookup.put("r12", 12); lookup.put("r13", 13); lookup.put("r14", 14); lookup.put("r15", 15);

		lookup.put("zr", 0); lookup.put("a0", 1); lookup.put("a1", 2); lookup.put("a2", 3);
		lookup.put("a3", 4); lookup.put("v0", 5); lookup.put("v1", 6); lookup.put("v2", 7);
		lookup.put("v3", 8); lookup.put("v4", 9); lookup.put("v5", 10); lookup.put("v6", 11);
		lookup.put("v7", 12); lookup.put("v8", 13); lookup.put("fp", 13); lookup.put("sp", 14);
		lookup.put("lr", 15);
	}
	
	public String getcode(int key) {
		return codes_inv.get(key);
	}
	
	public List<String> getsymbols() {
		List <String> symbols = new ArrayList<>();
		
		for (Map.Entry<String, Integer> entry : symtbl.entrySet()) {
			symbols.add(String.format("%04x", entry.getValue()) + " " + entry.getKey());
		}
						
		return symbols;
	}

	private static String tohex(int n) {
		return String.format("%04x", n & 0xffff);
	}

	private static int getval(String s) {
		if (s == null || s.isEmpty()) return 0;
		if (lookup.containsKey(s)) {
			return lookup.get(s);
		}
		try {
			return Integer.decode(s);
		} catch (NumberFormatException e) {
			throw new IllegalArgumentException("invalid symbol: " + s);
		}
	}

	public static void pass1(List<String> program) {
		for (int i = 0; i < program.size(); i++) {
			String line = program.get(i).trim();
			if (line.isEmpty() || line.startsWith(";")) {
				program.set(i, "");
				continue;
			}
			String[] fields = line.split("\\s+");
			if (fields.length > 0) {
				if (fields[0].equals("nop")) {
					program.set(i, "\tand\tr0,r0");
				} else if (fields[0].equals("hlt")) {
					program.set(i, "\thlt\tr0,r0");
				}
			}
		}
	}
	
	public static void pass2(List<String> program) {
		int pc = 0;
		for (String line : program) {
			String trimmed_line = line.trim();
			if (trimmed_line.isEmpty()) continue;

			String instruction_part = line;
			if (!Character.isWhitespace(line.charAt(0))) {
				String[] fields = trimmed_line.split("\\s+", 2);
				lookup.put(fields[0], pc);
				symtbl.put(fields[0], pc);
				instruction_part = (fields.length > 1) ? fields[1] : "";
			}

			instruction_part = instruction_part.trim();
			if (instruction_part.isEmpty()) continue;

			String[] fields = instruction_part.split("\\s+");
			
			if (codes.containsKey(fields[0])) {
				String all_operands = String.join("", Arrays.copyOfRange(fields, 1, fields.length));
				String[] parts_comment = all_operands.split(";");
				String[] parts = parts_comment[0].split(",");
				pc += (parts.length == 3) ? 4 : 2;
			} else {
				int quote_start = instruction_part.indexOf('"');
				int quote_end = instruction_part.lastIndexOf('"');
				int bracket_start = instruction_part.indexOf('[');
				int bracket_end = instruction_part.lastIndexOf(']');

				if (quote_start != -1 && quote_end > quote_start) {
					String content = instruction_part.substring(quote_start + 1, quote_end);
					content = content.replace("\\t", "\t");
					content = content.replace("\\n", "\n");
					content = content.replace("\\r", "\r");
					int len = content.length() + 1;
					if (len % 2 != 0)
						len++;
					pc += len;
				} else if (bracket_start != -1 && bracket_end > bracket_start) {
					String content = instruction_part.substring(bracket_start + 1, bracket_end);
					int len = Integer.decode(content);
					if (len % 2 != 0)
						len++;
					pc += len;
				} else {
					int dollar_start = instruction_part.indexOf('$');
					
					if (dollar_start != -1) {
						pc += fields.length;
						if (fields.length % 2 != 0) pc++;
					} else {
						pc += fields.length * 2;
					}
				}
			}
		}
	}

	private static long assemble(String[] fields) {
		String op = fields[0];
		Integer op_val = codes.get(op);

		if (op_val == null)
			return getval(op);
		
		String all_operands = String.join("", Arrays.copyOfRange(fields, 1, fields.length));
		String[] parts_comment = all_operands.split(";");
		String[] parts = parts_comment[0].split(",");
		
		if (parts.length == 2) {
			return op_val | (getval(parts[0]) << 4) | (getval(parts[1]) & 0xf);
		} else if (parts.length == 3) {
			long instruction_word = op_val | 0x0100 | (getval(parts[0]) << 4) | (getval(parts[1]) & 0xf);
			return (instruction_word << 16) | (getval(parts[2]) & 0xffff);
		}
		
		throw new IllegalArgumentException("incorrect number of operands: " + parts.length);
	}
	
	public static int pass3(List<String> program, List<String> output) {
		int pc = 0;
		int line_num = 0;

		for (String line : program) {
			line_num++;
			String trimmed_line = line.trim();
			if (trimmed_line.isEmpty()) continue;

			String instruction_part = line;
			if (!Character.isWhitespace(line.charAt(0))) {
				String[] fields = trimmed_line.split("\\s+", 2);
				instruction_part = (fields.length > 1) ? fields[1] : "";
			}
			
			instruction_part = instruction_part.trim();
			if (instruction_part.isEmpty()) continue;
			
			try {
				String[] fields = instruction_part.split("\\s+");

				if (codes.containsKey(fields[0])) {
					long assembled_code = assemble(fields);
					
					if (assembled_code > 0xffff) {
						int inst_word = (int)(assembled_code >>> 16);
						int imm_word = (int)(assembled_code & 0xffff);
						output.add(String.format("%s %s    (%s r%d,r%d,%d)",
								tohex(pc), tohex(inst_word), codes_inv.get(inst_word & 0xfe00),
								(inst_word & 0xf0) >> 4, (inst_word & 0xf), imm_word));
						pc += 2;
						output.add(String.format("%s %s", tohex(pc), tohex(imm_word)));
						pc += 2;
					} else {
						int instruction = (int)assembled_code;
						output.add(String.format("%s %s    (%s r%d,r%d)",
								tohex(pc), tohex(instruction), codes_inv.get(instruction & 0xfe00),
								(instruction & 0xf0) >> 4, (instruction & 0xf)));
						
						pc += 2;
					}
				} else {
					int quote_start = instruction_part.indexOf('"');
					int quote_end = instruction_part.lastIndexOf('"');
					int bracket_start = instruction_part.indexOf('[');
					int bracket_end = instruction_part.lastIndexOf(']');

					if (quote_start != -1 && quote_end > quote_start) {
						String content = instruction_part.substring(quote_start + 1, quote_end);
						content = content.replace("\\t", "\t");
						content = content.replace("\\n", "\n");
						content = content.replace("\\r", "\r");
						content = content + "\0";
						if (content.length() % 2 != 0)
							content = content + "\0";
						for (int i = 0; i < content.length(); i += 2) {
							int word = (content.charAt(i) << 8) | content.charAt(i + 1);
							output.add(String.format("%s %s", tohex(pc), tohex(word)));
							pc += 2;
						}
					} else if (bracket_start != -1 && bracket_end > bracket_start) {
						String content = instruction_part.substring(bracket_start + 1, bracket_end);
						int len = Integer.decode(content);
						if (len % 2 != 0)
							len++;
						for (int i = 0; i < len; i += 2) {
							output.add(String.format("%s 0000", tohex(pc)));
							pc += 2;
						}
					} else {
						int dollar_start = instruction_part.indexOf('$');
					
						if (dollar_start != -1) {
							ArrayList<String> fields_ar = new ArrayList<String>();
							for (String data_field : fields)
								fields_ar.add(data_field);
							if (fields.length % 2 != 0)
								fields_ar.add("$0");
							for (int f = 0; f < fields_ar.size(); f += 2) {
								String data_field1 = fields_ar.get(f).replace("$", "");
								String data_field2 = fields_ar.get(f + 1).replace("$", "");
								long data = (Integer.decode(data_field1) << 8) | Integer.decode(data_field2);
								output.add(String.format("%s %s", tohex(pc), tohex((int)data)));
								pc += 2;
							}
						} else {
							for (String data_field : fields) {
								long data = assemble(new String[]{data_field});
								output.add(String.format("%s %s", tohex(pc), tohex((int)data)));
								pc += 2;
							}
						}
					}
				}
			} catch (Exception e) {
				output.add(String.format("**** ????    assemble failed (%s)", e.getMessage()));
				output.add(String.format("             line %d --> \"%s\"", line_num, line.replace("\n", "")));
				return -1;
			}
		}
		return 0;
	}
}

File: python/assembler.py
#!/usr/bin/python3

import sys, string, re

codes = {
	"and":0x0000, "or":0x0400, "xor":0x0600, "add":0x0800,
	"sub":0x0a00, "lsl":0x3000, "lsr":0x3400, "asr":0x3600,
	"ldw":0x4000, "ldb":0x4400, "lbu":0x4600, "stw":0x4800,
	"stb":0x4c00, "blt":0x8000, "bge":0x8200, "bbl":0x8800,
	"bae":0x8a00, "beq":0x9000, "bne":0x9200, "tlt":0xa000,
	"tge":0xa200, "tbl":0xa800, "tae":0xaa00, "teq":0xb000,
	"tne":0xb200, "hlt":0xf800
}

lookup = {
	"r0":0, "r1":1, "r2":2, "r3":3,
	"r4":4, "r5":5, "r6":6, "r7":7,
	"r8":8, "r9":9, "r10":10, "r11":11,
	"r12":12, "r13":13, "r14":14, "r15":15,
	
	"zr":0, "a0":1, "a1":2, "a2":3,
	"a3":4, "v0":5, "v1":6, "v2":7,
	"v3":8, "v4":9, "v5":10, "v6":11,
	"v7":12, "v8":13, "fp": 13, "sp":14,
	"lr":15
}

def is_number(s):
	try:
		int(s)
		return True
	except ValueError:
		return False

def tohex(n):
	return "%s" % ("0000%x" % (n & 0xffff))[-4:]

def getval(s) :
	"return numeric value of a symbol or number"
	if not s : return 0							# empty symbol - zero
	a = lookup.get(s)							# get value or None if not in lookup
	if a == None : return int(s, 0)						# just a number (prefix can be 0x.. 0o.. 0b..)
	else : return a

def pass1(program) :
	"process pseudo operations"
	i = 0
	for lin in program :
		flds = str.split(lin)
		if flds :
			if flds[0] == ";" :
				program[i] = '\n'
			if flds[0] == "nop" :
				program[i] = "\tand	r0,r0\n"
			if flds[0] == "hlt" :
				program[i] = "\thlt	r0,r0\n"
		i += 1

def pass2(program) :
	"determine addresses for labels and add to the lookup dictionary"
	global lookup
	pc = 0
	for lin in program :
		flds = str.split(lin)
		if not flds : continue						# just an empty line
		if lin[0] > ' ' :
			symb = flds[0]						# a symbol - save its address in lookup
			lookup[symb] = pc
		else :
			if codes.get(flds[0]) == None :
				flds2 = ' '.join(flds)
				if flds2 :
					if flds2[0] == '"' and flds2[-1] == '"' :
						flds2 = lin
						flds2 = flds2[1:-1]
						flds2 = flds2.replace("\\t", chr(0x09))
						flds2 = flds2.replace("\\n", chr(0x0a))
						flds2 = flds2.replace("\\r", chr(0x0d))
						while (flds2[0] != '"') :
							flds2 = flds2[1:]
						flds2 = flds2[1:-1] + '\0'
						while (len(flds2) % 2) != 0 :
							flds2 = flds2 + '\0'
						pc = pc + len(flds2)
					elif flds2[0] == '[' and flds2[-1] == ']' :
						flds2 = lin
						flds2 = int(flds2[2:-2], 0)
						if (flds2 % 2) != 0 :
							flds2 = flds2 + 1
						pc = pc + flds2
					else:
						if flds2[0] == '$' :
							for f in flds :
								pc = pc + 1
								
							if (len(flds) % 2) != 0 :
								pc = pc + 1
						else :
							for f in flds :
								pc = pc + 2
			else :
				parts = ''.join(flds[1:])
				parts = str.split(parts, ";");					# strip comments on code lines
				parts = parts[0]
				parts = str.split(parts, ",")					# break opcode fields
				if len(parts) == 2 :
					pc = pc + 2
				if len(parts) == 3 :
					pc = pc + 4

def assemble(flds) :
	"assemble instruction to machine code"
	opval = codes.get(flds[0])
	symb = lookup.get(flds[0])
	if symb != None :
		return symb
	else :
		if opval == None : return int(flds[0], 0)			# just a number (prefix can be 0x.. 0o.. 0b..)
		parts = ''.join(flds[1:])
		parts = str.split(parts, ";");					# strip comments on code lines
		parts = parts[0]
		parts = str.split(parts, ",")					# break opcode fields
		if len(parts) == 2 :
			parts = [0, parts[0], parts[1]]
			return (opval | (getval(parts[1]) << 4) | (getval(parts[2]) & 0xf))
		if len(parts) == 3 :
			parts = [0, parts[0], parts[1], parts[2]]
			return (((opval | 0x0100 |
				(getval(parts[1]) << 4) | 
				(getval(parts[2]) & 0xf)) << 16) |
				getval(parts[3]) & 0xffff)

def pass3(program) :
	"translate assembly code and symbols to machine code"
	
	codes_inv = {v: k for k, v in codes.items()}
	
	args = sys.argv[1:]
	if args :
		args = args[0]
	else :
		args = ''

	pc = 0
	line = 0

	for lin in program :
		line = line + 1
		flds = str.split(lin)
		if lin[0] > ' ' : flds = flds[1:]			# drop symbol if there is one
		if not flds : continue
		try :
			flds2 = ' '.join(flds)
			if flds2[0] == '"' and flds2[-1] == '"' :
				flds2 = lin
				flds2 = flds2[1:-1]
				flds2 = flds2.replace("\\t", chr(0x09))
				flds2 = flds2.replace("\\n", chr(0x0a))
				flds2 = flds2.replace("\\r", chr(0x0d))
				while (flds2[0] != '"') :
					flds2 = flds2[1:]
				flds2 = flds2[1:-1] + '\0'
				while (len(flds2) % 2) != 0 :
					flds2 = flds2 + '\0'
				flds3 = ''
				while True :
					flds3 += (str((int(ord(flds2[0])) << 8) |
						int(ord(flds2[1])))) + ' '
					flds2 = flds2[2:]
					if flds2 == '' : break
				flds3 = str.split(flds3)
				instruction = assemble(flds3)
				print ("%04x %s" % (pc, tohex(instruction)))
				pc = pc + 2
				flds3 = flds3[1:]
				for f in flds3 :
					instruction = assemble(flds3)
					print ("%04x %s" % (pc, tohex(instruction)))
					pc = pc + 2
					flds3 = flds3[1:]
				flds = ''
			elif flds2[0] == '[' and flds2[-1] == ']' :
				flds2 = lin
				flds2 = int(flds2[2:-2], 0)
				if (flds2 % 2) != 0 :
					flds2 = flds2 + 1
				for i in range(0, flds2, 2) :
					print ("%04x 0000" % pc)
					pc = pc + 2
			else :
				if codes.get(flds[0]) == None :
					flds2 = ' '.join(flds)
					if (flds2[0] == '$') :
						if (len(flds) % 2) != 0 :
							flds.append('$0')
							
						for f in flds :
							if (len(flds) % 2) == 0 : 
								data = ((int(flds[0][1:], 0) & 0xff) << 8) | (int(flds[1][1:], 0) & 0xff)
								print ("%04x %s" % (pc, tohex(data)))
								pc = pc + 2
								flds = flds[1:]
							else :
								flds = flds[1:]
								continue
					else :
						for f in flds :
							data = assemble(flds)
							print ("%04x %s" % (pc, tohex(data)))
							pc = pc + 2
							flds = flds[1:]
				else :
					instruction = assemble(flds)
					if instruction < 65536 :	# instruction without immediate field
						print ("%04x %s    (%s r%d,r%d)" % (pc, tohex(instruction),
							codes_inv[instruction & 0xfe00],
							(instruction & 0xf0) >> 4,
							(instruction & 0xf)))
						pc = pc + 2
					else :				# instruction with immediate field
						inst = instruction >> 16
						imm = instruction & 0xffff
						print ("%04x %s    (%s r%d,r%d,%d)" % (pc, tohex(inst),
							codes_inv[inst & 0xfe00],
							(inst & 0xf0) >> 4,
							(inst & 0xf), imm))
						pc = pc + 2
						print ("%04x %s" % (pc, tohex(imm)))
						pc = pc + 2
						
		except :
			print ("**** ????    line %d --> %s" % (line, lin.replace(chr(0x0a), ' ')))
			return -1
	return 0

def main() :
	sys.stdin.reconfigure(errors = 'ignore')
	program = sys.stdin.readlines()
	pass1(program)
	pass2(program)
	pass3(program)

if __name__ == "__main__" : main()
